### PR TITLE
Rework loop fuzzy logic for NumberConstraints

### LINKED ISSUE #8028: Fix S2583 FP: Loop with manually incremented counter
### Description

S2583 raises a warning for the following code:

```cs
public void Test(int[] items)
{
    var m = 0;
    foreach (var item in items)
    {
        m++;
        if (m % 3 == 0) // Noncompliant - FP
        {
            // do something
        }
    }
}
```

**Note:** the warning is not raised when 2 is used instead of 3 in the modulo operation:
```cs
...
if (m % 2 == 0) // Does not raise
...
```

### Expected behavior

The analyzer should not raise an issue.

### Actual behavior

The analyzer raises an issue: `S2583: Change this condition so that it does not always evaluate to 'False'.`

### Related information

* C#/VB.NET Plugins version: 9.10.0.77988
* Visual Studio version: Visual Studio 2022 (version 17.7)
* MSBuild / dotnet version: .NET 7.0
* Operating System: Windows 10

### LINKED ISSUE #8428: Fix S2583/S2589 FP: For loop with Array.Length
The SE engine does not learn `Array.Length` which results in FPs.
This could be solved by tracking array sizes or by changing how the engine handles for-loops with unknown limits.

```CS
void Test()
{
    var ids = new int[1000];
    for (var i = 0; i < ids.Length; i++)
    {
        if (i % 100 != 0 || i <= 0)   // Noncompliant FP
        {
            System.Diagnostics.Debug.WriteLine(i);
        }
    }
}
```

Similar problems arise with C#12 [collection expressions](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#collection-expressions):

```CS
void CollectionExpressions(int[] array)
{
    int[] knownLength = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    for (var i = 0; i < knownLength.Length; i++)
    {
        if (i > 5)                  // Noncompliant FP
        {
            Console.WriteLine(i);   // Secondary
        }
    }

    int[] unknownLength = [1, 2, .. array, 3, 4];
    for (var i = 0; i < unknownLength.Length; i++)
    {
        if (i > 5)                  // Noncompliant FP
        {
            Console.WriteLine(i);   // Secondary
        }
    }

    int[] knownLength2 = [1, 2, .. knownLength, 3, 4];
    for (var i = 0; i < knownLength.Length; i++)
    {
        if (i > 5)                  // Noncompliant FP
        {
            Console.WriteLine(i);   // Secondary
        }
    }
}
```

and [inline arrays](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#inline-arrays):
```cs
[System.Runtime.CompilerServices.InlineArray(10)]
public struct Buffer
{
    private int _element0;
}
var buffer = new Buffer(); // buffer.Length is 10
```

Originally reported [here](https://community.sonarsource.com/t/another-false-positive-for-s2583-conditionally-executed-code-should-be-reachable/104561).

### LINKED ISSUE #8449: Fix S2589 FP: Change this condition so that it does not always evaluate to 'True'
### Description

The code below erroneously reports 2589 Change this condition so that it does not always evaluate to 'True' on the expression ```delay < 409600```

### Repro steps

````C#
public static void Foo()
{
    int delay = 100;
    while (true)
    {
        if (delay < 409600) delay = delay * 2;
    }
}
````
### Expected behavior

No warning

### Actual behavior

Warning S2589

### Known workarounds

None

### Related information

* C#/VB.NET Plugins version 9.15.0.81779
* Visual Studio version 17.7
* MSBuild / dotnet version 7
* Operating System Windows 11

### LINKED ISSUE #8474: Rework loop handling to also work for counter variables that are not part of the loop condition
[Plan](https://docs.google.com/document/d/1NRCDaS3M4mTBMJow_7jAKk-0VPWfNX57EO5rFqEULsw/edit?usp=sharing)
PoC: #8452
should fix: #8028, #8428, #8449

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixes #8474 
Fixes #8028
Fixes #8428
Fixes #8449

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs [hunk 1]
```diff
@@ -2271,7 +2271,7 @@ public void FP_Increment(List<int> list)
             var steps = 0;
             while (list.Any())
             {
-                if (steps++ > MaxStepCount) // Noncompliant FP
+                if (steps++ > MaxStepCount) // Compliant
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:2274]: Let's add test cases on:
- boundary conditions
- nested loops

They will prove useful when the loop handling is further improved in the future.
e.g.

```cs
void ForLoop_Regular()
{
    for (int i = 0; i < 10; i++)
    {
        _ = i > -1 ? 0 : 42;        // Noncompliant
        _ = i >= -1 ? 0 : 42;       // Noncompliant
        _ = i > 0 ? 0 : 42;         // Compliant
        _ = i >= 0 ? 0 : 42;        // Noncompliant
        _ = i > 9 ? 0 : 42;         // Noncompliant
        _ = i >= 9 ? 0 : 42;        // Compliant
        _ = i > 10 ? 0 : 42;        // Noncompliant
        _ = i >= 10 ? 0 : 42;       // Noncompliant
        _ = i >= 10 ? 0 : 42;       // Noncompliant
        _ = i >= i + 1 ? 0 : 42;    // FN
    }
}

void ForLoop_IncreaseInside()
{
    for (int i = 0; i < 10;)
    {
        _ = i > -1 ? 0 : 42;        // Noncompliant
        _ = i >= -1 ? 0 : 42;       // Noncompliant
        _ = i > 0 ? 0 : 42;         // Compliant
        _ = i >= 0 ? 0 : 42;        // Noncompliant
        _ = i > 9 ? 0 : 42;         // Noncompliant
        _ = i >= 9 ? 0 : 42;        // Compliant
        _ = i > 10 ? 0 : 42;        // Noncompliant
        _ = i >= 10 ? 0 : 42;       // Noncompliant
        _ = i >= 10 ? 0 : 42;       // Noncompliant
        _ = i >= i + 1 ? 0 : 42;    // FN
        i++;
    }
}

void ForLoop_Nested()
{
    for (int i = 0; i < 10; i++)
        for (int j = i + 1; j < i + 10; j++)
        {
            _ = j >= 0 ? 0 : 42;        // Noncompliant
            _ = j > 0 ? 0 : 42;         // Noncompliant
            _ = j < 18 ? 0 : 42;        // Compliant
            _ = j < 19 ? 0 : 42;        // Noncompliant
            _ = i + j < 17 ? 0 : 42;    // Compliant
            _ = i + j < 28 ? 0 : 42;    // Noncompliant
            _ = i != j ? 0 : 42;        // FN
            _ = i > j ? 0 : 42;         // FN
        }
}

void ForLoop_Nested_IncrementInside()
{
    for (int i = 0; i < 10;)
    {
        for (int j = i + 1; j < 10;)
        {
            _ = j >= 0 ? 0 : 42;        // Noncompliant
            _ = j > 0 ? 0 : 42;         // Noncompliant
            _ = j < 18 ? 0 : 42;        // Compliant
            _ = j < 19 ? 0 : 42;        // Noncompliant
            _ = i + j < 17 ? 0 : 42;    // Compliant
            _ = i + j < 28 ? 0 : 42;    // Noncompliant
            _ = i != j ? 0 : 42;        // FN
            _ = i > j ? 0 : 42;         // FN
            j++;
        }
        i++;
    }
}

void ForLoop_MultipleLoopVariables()
{
    for (int i = 0, j = 1; i < 10; i++, j++)
    {
        _ = i >= 0 ? 0 : 42;            // Noncompliant
        _ = j > 0 ? 0 : 42;             // Noncompliant
        _ = j < 3 ? 0 : 42;             // Noncompliant - FP
        _ = i + j > 0 ? 0 : 42;         // Noncompliant
        _ = i + j > 1 ? 0 : 42;         // Compliant
    }
}

void ForLoop_ZeroOrOneExecutions()
{
    for (int i = 0; i < 0; i++)
    {
        _ = i > 0 ? 0 : 42;             // Compliant
    }

    for (int i = 0; i < 1; i++)
    {
        _ = i > 0 ? 0 : 42;             // Noncompliant
    }

    for (int i = 0; i < 1; ++i)
    {
        _ = i > 0 ? 0 : 42;             // Noncompliant
    }
}
```

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM

File: analyzers/its/expected/Ember-MM/generic.EmberCore.XBMC-{6FE33C61-E2C2-4982-9536-63AEF0A98AAA}-S2583.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/its/expected/Ember-MM/generic.EmberCore.XBMC-{6FE33C61-E2C2-4982-9536-63AEF0A98AAA}-S2583.json:1]: Yes, this is a limitation with the loop detection: The code that changes `retry` is inside the `catch`-block which is not correctly detected as being inside the loop => loop fuzzy logic is not applied.

[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/its/expected/Ember-MM/generic.EmberCore.XBMC-{6FE33C61-E2C2-4982-9536-63AEF0A98AAA}-S2583.json:1]: Yes, this is a limitation with the loop detection: The code that changes `retry` is inside the `catch`-block which is not correctly detected as being inside the loop => loop fuzzy logic is not applied.
Therefore the exception branches cannot leave the loop, because `retry > 0` is always true (for the first two iterations).

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs (hunk 1 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:2274]: Very nice test cases, thank you!
